src/commands.c

```diff
@@ -236,59 +236,98 @@ static void freeCommandSpecDict(RedisModuleDict *dict)
     RedisModule_FreeDict(NULL, dict);
 }
 
-RRStatus CommandSpecSet(RedisModuleCtx *ctx, const char *ignored_commands)
+RRStatus CommandSpecInit(RedisModuleCtx *ctx)
 {
-    RedisModuleDict *dict = RedisModule_CreateDict(ctx);
+    RedisModule_Assert(commandSpecDict == NULL);
+    commandSpecDict = RedisModule_CreateDict(NULL);
 
     for (int i = 0; commands[i].name != NULL; i++) {
         CommandSpec *cs = RedisModule_Alloc(sizeof(*cs));
         cs->name = RedisModule_Strdup(commands[i].name);
         cs->flags = commands[i].flags;
 
-        int ret = RedisModule_DictSetC(dict, cs->name, strlen(cs->name), cs);
+        int ret = RedisModule_DictSetC(commandSpecDict, cs->name, strlen(cs->name), cs);
         if (ret != REDISMODULE_OK) {
             goto error;
         }
     }
 
-    if (ignored_commands) {
-        char *tok, *s = NULL;
-        char *tmp = RedisModule_Strdup(ignored_commands);
+    populateCommandSpecFromRedis(ctx);
 
-        for (tok = strtok_r(tmp, ",", &s); tok; tok = strtok_r(NULL, ",", &s)) {
-            int nokey = 0;
-            CommandSpec *cs;
+    return RR_OK;
 
-            cs = RedisModule_DictGetC(dict, tok, strlen(tok), &nokey);
-            if (!nokey) {
-                cs->flags = CMD_SPEC_DONT_INTERCEPT;
-                continue;
-            }
+error:
+    freeCommandSpecDict(commandSpecDict);
+    commandSpecDict = NULL;
+    return RR_ERROR;
+}
 
-            cs = RedisModule_Calloc(1, sizeof(*cs));
-            cs->name = RedisModule_Strdup(tok);
-            cs->flags = CMD_SPEC_DONT_INTERCEPT;
+RRStatus CommandSpecFree()
+{
+    freeCommandSpecDict(commandSpecDict);
+    return RR_OK;
+}
+
+static RRStatus updateIgnoredCommands(RedisModuleDict *command_spec_dict, const char *commands_str, bool ignore)
+{
+    if (!commands_str) {
+        return RR_OK;
+    }
+
+    char *tok, *s = NULL;
+    char *tmp = RedisModule_Strdup(commands_str);
 
-            int ret = RedisModule_DictSetC(dict, tok, strlen(tok), cs);
-            if (ret != REDISMODULE_OK) {
-                RedisModule_Free(cs->name);
-                RedisModule_Free(cs);
-                RedisModule_Free(tmp);
-                goto error;
+    for (tok = strtok_r(tmp, ",", &s); tok; tok = strtok_r(NULL, ",", &s)) {
+        int nokey = 0;
+        CommandSpec *cs;
+
+        cs = RedisModule_DictGetC(command_spec_dict, tok, strlen(tok), &nokey);
+        RedisModule_Assert(!nokey || ignore);
+        if (!nokey) {
+            if (!ignore) {
+                cs->flags &= ~CMD_SPEC_DONT_INTERCEPT;
+            } else {
+                cs->flags |= CMD_SPEC_DONT_INTERCEPT;
             }
+            continue;
+        }
+
+        cs = RedisModule_Calloc(1, sizeof(*cs));
+        cs->name = RedisModule_Strdup(tok);
+        cs->flags = CMD_SPEC_DONT_INTERCEPT;
+
+        int ret = RedisModule_DictSetC(command_spec_dict, tok, strlen(tok), cs);
+        if (ret != REDISMODULE_OK) {
+            RedisModule_Free(cs->name);
```

Comment:
Previously, we were rebuilding a separate the map on each config change and then replacing it atomically if we are successful. So, it had a few advantages. 
- If user passes a bad config, we wouldn't have partial configs. With this PR, if user passes `ignored_commands`, say `cmd1, cmd2, cmd1, cmd3`, we are going to add `cmd1,cmd2` to the map and also return an error. (see the duplicate command). 
- Currently, we don't delete entries from the map. If `ignored` commands are `cmd1,cmd2` and new one is empty string, entries stay in the map. 
- Similar problem for module commands. If we unload a module, we will not delete its commands from the dictionary. Previously, that wasn't the case as we create a new dict each time. 

Overall, I'm not sure what is the benefit of different function calls for setting the ignored commands and other commands. I guess rebuilding the dict fully is simpler in our case. 

----
If this is the way we should do this (not sure), maybe we should have `CMD_SPEC_IGNORED` flag and mark user specified configs as `CMD_SPEC_IGNORED | CMD_SPEC_DONT_INTERCEPT`. 

Then, if user wants to update `ignored-commands`, we can create a temp dict, copy `non - CMD_SPEC_IGNORED` ones there, create new cmdspecs for the new `ignored-commands` and then atomically replace the map.

Similar for redis commands, on update, we copy `CMD_SPEC_IGNORED` ones to a new map and then fetch new redis commands to this map, an then atomically replace the map.

Reply:
1. regarding the "bad config" - actually even in the orig implementation we ignored duplication (see https://github.com/RedisLabs/redisraft/blob/72b4741a1fa30b28ce800749c77e59bf29c66984/src/commands.c#L263). The difference is in case of error (oom): here we add part of the cmds to the dict, instead releasing the whole dict, and I think we should keep this behaviour. 
2. Right, there is an assumption here that `ignored_commands` should be valid commands only, which is part of the command dict anyway. I agree we have several issues here: 
- Does empty "ignored-commands" means "removing the previous restrictions"? - if yes we should support it (it's during config time, so no performance concerns)
- Agree that for module commands we have no other way but to update the whole dict.

Considering the above issues, we should update the dict each time - will update my code accordingly.

Reply:
1- Just to be clear, in case of an error, we dismiss the temporary dict, so we don't store the new/bad config partially. 

2- 
> Does empty "ignored-commands" means "removing the previous restrictions"? - if yes we should support it (it's during config time, so no performance concerns)

Yes. 

Again just to be clear, what I meant is , with the current PR, we don't ever delete commanspecs from the dict. Let's say previous config is `cmd1,cmd2` and new config is `cmd3,cmd4`. With this PR, `cmd1,cmd2` will stay in the dict? This wasn't the case previously. 

----

Overall, I feel like we just need a single function rebuild the dict each time. (We already have it, `CommandSpecSet()` )
I guess we can rename  `CommandSpecSet()` as `CommandSpecRebuild()` for clarity, maybe. 

Because if `CommandSpecInit()`, `CommandSpecUpdateIgnoredCommands()` and `CommandSpecUpdateRedisCommands()` will do the same thing, rebuild the dict, we don't need three functions. At least, this was my conclusion. Maybe, I'm wrong. 

I know it is ugly currently as we call this in the config path but don't know if there is a better way to do it.

Reply:
For sure we don't want to call it *only* in the config path, because `ignored-commands` is not a mandatory configuration, meaning command spec will not be created at all. We should have a single function, which we call on 3 different paths - onLoad, on config change and on module load state change.

Reply:
> 1- Just to be clear, in case of an error, we dismiss the temporary dict, so we don't store the new/bad config partially.

Right, but in the current implementation duplication is not an error. 

Anyway, it seems like we treat differently to a `RedisModule_DictSetC()` error - in some cases we assert on that (see https://github.com/RedisLabs/redisraft/blob/72b4741a1fa30b28ce800749c77e59bf29c66984/src/commands.c#L128) on other cases, like in `CommandSpecSet` (https://github.com/RedisLabs/redisraft/blob/72b4741a1fa30b28ce800749c77e59bf29c66984/src/commands.c#L248), we return an error, and ignoring... we should be consistent...

Reply:
> For sure we don't want to call it only in the config path, because ignored-commands is not a mandatory configuration, meaning command spec will not be created at all

No, it will be called for the default value even if the config is missing. 

> Right, but in the current implementation duplication is not an error.

Ahh, I see, sorry. Yeah, duplication is not an error, just image another error case instead :) 

> Anyway, it seems like we treat differently to a RedisModule_DictSetC() error - in some cases we assert on that 

Depends if we can handle the error not I guess.